libp2p/protocols/discovery/rendezvous.nim

```diff
@@ -0,0 +1,348 @@
+                               text: string = "") {.async.} =
+  let msg = Protobuf.encode(Message(
+        msgType: some(MessageType.DiscoverResponse),
+        discoverResponse: some(DiscoverResponse(status: some(status), text: some(text)))))
+  await conn.writeLp(msg)
+
+proc countRegister(rdv: RendezVous, peerId: PeerId): int =
+   for data in rdv.registered:
+     if data.peerId == peerId:
+       result.inc()
+
+proc save(rdv: RendezVous, ns: string, peerId: PeerId, r: Register, ttl: uint64 = MinimumTTL) =
+  let
+    n = now()
+    nsSalted = ns & rdv.salt
+  discard rdv.indexes.hasKeyOrPut(nsSalted, initOffsettedSeq[int]())
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == peerId:
+        rdv.registered[index].expireDate = rdv.defaultDT
+    rdv.registered.add(
+      RegisteredData(
+        peerId: peerId,
+        expireDate: now() + initDuration(ttl.int64),
+        data: r
+      )
+    )
+    rdv.indexes[nsSalted].add(rdv.registered.high)
+#    rdv.registerEvent.fire()
+  except KeyError:
+    doAssert false, "Should have key"
+
+proc register(rdv: RendezVous, conn: Connection, r: Register): Future[void] =
+  let ns = r.ns.get("")
+  if ns.len notin 1..255:
+    return conn.sendRegisterResponseError(InvalidNamespace)
+  let ttl = r.ttl.get(MinimumTTL)
+  if ttl notin MinimumTTL..MaximumTTL:
+    return conn.sendRegisterResponseError(InvalidTTL)
+  let pr = checkPeerRecord(r.signedPeerRecord, conn.peerId)
+  if pr.isErr():
+    return conn.sendRegisterResponseError(InvalidSignedPeerRecord, pr.error())
+  if rdv.countRegister(conn.peerId) >= RegistrationLimitPerPeer:
+    return conn.sendRegisterResponseError(NotAuthorized, "Registration limit reached")
+  rdv.save(ns, conn.peerId, r, ttl)
+  conn.sendRegisterResponse(ttl)
+
+proc unregister(rdv: RendezVous, conn: Connection, u: Unregister) =
+  let nsSalted = u.ns.get("") & rdv.salt
+  try:
+    for index in rdv.indexes[nsSalted]:
+      if rdv.registered[index].peerId == conn.peerId:
+        rdv.registered[index].expireDate = rdv.defaultDT
+  except KeyError:
+    return
+
+proc discover(rdv: RendezVous, conn: Connection, d: Discover): Future[void] =
+  let ns = d.ns.get("")
+  if ns.len notin 0..255:
+    return conn.sendDiscoverResponseError(InvalidNamespace)
+  var
+    limit = min(DiscoverLimit, d.limit.get(DiscoverLimit))
+  let
+    nsSalted = ns & rdv.salt
+    cookie = if d.cookie.isSome():
+      try: Protobuf.decode(d.cookie.get(), type(Cookie))
+      except ProtobufReadError: return conn.sendDiscoverResponseError(InvalidCookie)
+    else: Cookie(offset: rdv.registered.low().uint64, ns: ns)
+    indexes = if ns != "":
+        try: rdv.indexes[nsSalted].s
+        except KeyError: return conn.sendDiscoverResponseError(InvalidNamespace)
+      else: toSeq(rdv.registered.low()..rdv.registered.high())
+  if indexes.len() == 0:
+    return conn.sendDiscoverResponse(@[], Cookie())
```

Comment:
Shouldn't we send a None Cookie in this case? Not sure it's worth the trouble though

Reply:
It's the same thing